.github/workflows/e2e.yaml

```diff
@@ -0,0 +1,222 @@
+                --header "x-attribution-login: ${{ github.actor }}" --header "x-attribution-actor-id: ${{ github.actor }}" \
+                --header "Circle-Token: ${{ secrets.circleCIToken }}" $DATA )
+
+              if [[ "$STATUS_CODE" == "429" ]]; then
+                STATUS_CODE=
+                echo '' > $OUTPUT_FILE
+                sleep $((RANDOM % 3))
+              fi
+            done
+
+            cat $OUTPUT_FILE
+            rm $OUTPUT_FILE
+            if [ $STATUS_CODE -lt 200 ] || [ $STATUS_CODE -gt 399 ] ; then
+              echo "Error requesting $METHOD $URL (status $STATUS_CODE)"
+              exit 1
+            fi
+          }
+
+          PIPELINE_ID=$(request POST $CIRCLE_CI_API_PATH "$BODY" | jq -r '.id')
+          sleep 3
+          WORKFLOW_DETAILS=$(request GET https://circleci.com/api/v2/pipeline/$PIPELINE_ID/workflow | jq -cr '.items[] | select(.name == "manual-e2e")')
+          PIPELINE_NUMBER=$(echo $WORKFLOW_DETAILS | jq -r '.pipeline_number')
+          WORKFLOW_ID=$(echo $WORKFLOW_DETAILS | jq -r '.id')
+
+          echo "pipeline_number=$PIPELINE_NUMBER" >> $GITHUB_OUTPUT
+          echo "workflow_id=$WORKFLOW_ID" >> $GITHUB_OUTPUT
+
+          if [[ "$WORKFLOW_ID" == "" ]]; then
+            echo "Could not trigger a workflow on CircleCI, check your .circleci/config.yaml"
+            exit 1
+          fi
+
+          echo ''
+          echo "CircleCI pipeline triggered successfully, pipeline id: $PIPELINE_ID"
+          echo "Check CircleCI workflow details at: https://app.circleci.com/pipelines/gh/${{ github.repository }}/$PIPELINE_NUMBER/workflows/$WORKFLOW_ID"
+      - name: "Check run status of pipeline workflow on CircleCI"
+        if: steps.circle-ci-trigger.outputs.workflow_id != ''
+        run: |
+          set -e
+          if [ "${{ runner.debug }}" == "1" ]; then
+            set -x
+          fi
+
+          function request(){
+            METHOD=$1
+            URL=$2
+            DATA=$3
+
+            if [ "$DATA" != "" ]; then
+              DATA="--data $DATA"
+            fi
+
+            OUTPUT_FILE=/tmp/circleci-response-$RANDOM.json
+            STATUS_CODE=$(curl -o $OUTPUT_FILE -sL -w "%{http_code}" -X $METHOD $URL \
+              --header "content-type: application/json" --header "accept: application/json" \
+              --header "x-attribution-login: ${{ github.actor }}" --header "x-attribution-actor-id: ${{ github.actor }}" \
+              --header "Circle-Token: ${{ secrets.circleCIToken }}" $DATA )
+
+            cat $OUTPUT_FILE
+            rm $OUTPUT_FILE
+            if [ "$STATUS_CODE" == "429" ]; then
+              # we are exceeding rate limit
+              echo "{}"
+              return
+            fi
+            if [ $STATUS_CODE -lt 200 ] || [ $STATUS_CODE -gt 399 ] ; then
+              echo "Error requesting $METHOD $URL (status $STATUS_CODE)"
+              exit 1
+            fi
+          }
+
+          function check_workflow(){
```

Comment:
Also a lot of big inline shell scripts maybe better in separate files?

Reply:
The structure of the script is quite simple though.